lib/actions/deploy.coffee

```diff
@@ -1,229 +1,183 @@
 			boolean: true
 		}
 	]
 	action: (params, options, done) ->
-		_ = require('lodash')
-		tmp = require('tmp')
-		tmpNameAsync = Promise.promisify(tmp.tmpName)
-		resin = require('resin-sdk-preconfigured')
-
+		helpers = require('../utils/helpers')
 		Logger = require('../utils/logger')
+
 		logger = new Logger()
 
-		# Ensure the tmp files gets deleted
-		tmp.setGracefulCleanup()
-
-		logs = ''
-
-		upload = (token, username, url) ->
-			dockerUtils.getDocker(options)
-			.then (docker) ->
-				# Check input parameters
-				parseInput(params, options)
-				.then ([appName, build, source, imageName]) ->
-					tmpNameAsync()
-					.then (bufferFile) ->
-
-						# Setup the build args for how the build routine expects them
-						options = _.assign({}, options, { appName })
-						params = _.assign({}, params, { source })
-
-						Promise.try ->
-							if build
-								dockerUtils.runBuild(params, options, getBundleInfo, logger)
-							else
-								{ image: imageName, log: '' }
-						.then ({ image: imageName, log: buildLogs }) ->
-							logger.logInfo('Initializing deploy...')
-
-							logs = buildLogs
-							Promise.all [
-								dockerUtils.bufferImage(docker, imageName, bufferFile)
-								token
-								username
-								url
-								params.appName
-								logger
-							]
-							.spread(performUpload)
-						.finally ->
-							# If the file was never written to (for instance because an error
-							# has occured before any data was written) this call will throw an
-							# ugly error, just suppress it
-							Promise.try ->
-								require('mz/fs').unlink(bufferFile)
-							.catch(_.noop)
-				.tap ({ image: imageName, buildId }) ->
-					logger.logSuccess("Successfully deployed image: #{formatImageName(imageName)}")
-					return buildId
-				.then ({ image: imageName, buildId }) ->
-					if logs is '' or options.nologupload?
-						return ''
-
-					logger.logInfo('Uploading logs to dashboard...')
-
-					Promise.join(
-						logs
-						token
-						url
-						buildId
-						username
-						params.appName
-						uploadLogs
-					)
-					.return('Successfully uploaded logs')
-				.then (msg) ->
-					logger.logSuccess(msg) if msg isnt ''
-				.asCallback(done)
-
-		Promise.join(
-			resin.auth.getToken()
-			resin.auth.whoami()
-			resin.settings.get('resinUrl')
-			upload
-		)
+		logger.logDebug('Parsing input...')
+
+		Promise.try ->
+			{ appName, image } = params
+
+			# look into "resin build" options if appName isn't given
+			appName = options.application if not appName?
+			delete options.application
+
+			if not appName?
+				throw new Error('Need an application to deploy to!')
+
+			if image? and options.build
+				throw new Error('Build option is not applicable when specifying an image')
+
+			Promise.join(
+				helpers.getApplication(appName)
+				helpers.getArchAndDeviceType(appName)
+				(app, { arch, device_type }) ->
+					app.arch = arch
+					app.device_type = device_type
+					return app
```

Comment:
NIT: these three lines could be a single `_.assign(app, { arch, device_type })`.

Reply:
It could, but I'd have to import lodash and I think there's no need for that (not that it would save anything, because it's already imported elsewhere, but "less imports > more imports" IMO :)